tutorial/lesson_24_async.cpp

```diff
@@ -0,0 +1,198 @@
+        // to wait until the data is fully ready. Wouldn't it be great if we could overlap computations
+        // of `producer` and `consumer` too? This computational pattern is known as 'double buffering' and
+        // can be critical for achieving good performance in certain scenarios. The high-level idea is that
+        // producer is allowed to run ahead and do the next chunk of work without waiting while consumer 
+        // is processing the current chunk. The obvious drawback of this method is that it requires twice
+        // as much memory for `producer`. 
+        Func producer("producer"), consumer("consumer");
+
+        producer(x, y, c) = (x + y) * (c + 1);
+        consumer(x, y, c) = producer(x - 1, y - 1, c) + producer(x, y, c) + producer(x + 1, y + 1, c);
+
+        consumer.compute_root();
+
+        // In this example the planes are processed separately, so producer can run ahead 
+        // and start producing plane `c + 1`, while `consumer` consumes already produced plane `c`.
+        // One way to express it with Halide schedule is very similar to how sliding window schedules
+        // are expressed (see lesson_8 for details). There are indeed a lot of commonalities between two
+        // because both of them are relying on a circular buffer as underlying data structure.
+        producer.compute_root()
+            .async()
+            .compute_at(consumer, c)
+            .store_at(consumer, Var::outermost())
+            // Explicit fold_storage is required here, because otherwise Halide will infer that only
+            // one plane of `producer` is necessary for `consumer`, but for the purposes of this
+            // example we want at least 2.
+            .fold_storage(c, 2);
+
+        consumer.realize({128, 128, 4});
+    }
+
+    {
+        // In the previous example, we relied on storage folding to to use double buffering
+        // technique, but there is another, more direct way to express that.
+        Func producer("producer"), consumer("consumer");
+
+        producer(x, y, c) = (x + y) * (c + 1);
+        consumer(x, y, c) = producer(x - 1, y - 1, c) + producer(x, y, c) + producer(x + 1, y + 1, c);
+
+        consumer.compute_root();
+
+        // As mentioned in the previous example, the planes are processed separately, so producer can run
+        // ahead and start producing plane `c + 1`, while `consumer` consumes already produced plane `c`.
+        // A more direct way to express this would be to hoist storage of `producer` to ouside of the loop
+        // `c` over planes, double its size and add necessary indices to flip between two planes.
+        // The first part can be achieved with `hoist_storage` directive and the rest is done with 
+        // `double_buffer`.
+        producer.compute_root()
+            .async()
+            .compute_at(consumer, c)
+            .hoist_storage(consumer, Var::outermost())
+            .double_buffer()
+            .async();
+
```

Comment:
I've added a detailed "high-level structure of the generated code" for the previous example and for this one it will be very similar (left a comment about it).

> Is hoist_storage() + double_buffer() + async() not conceptually the same as fold_storage() + async()?

They are the same in this specific case, but the next example demonstrates the difference (basically, impossible to do with fold_storage).

Reply:
I've added a detailed "high-level structure of the generated code" for the previous example and for this one it will be very similar (left a comment about it).

> Is hoist_storage() + double_buffer() + async() not conceptually the same as fold_storage() + async()?

They are the same in this specific case, but the next example demonstrates the difference (basically, impossible to do with fold_storage).